File: android/idsm_manager/app/src/main/java/com/idsm/manager/PropertyReport.kt
package com.idsm.manager

import org.json.JSONArray
import org.json.JSONObject
import kotlin.random.Random

/**
 * PropertyReport -- 设备属性/心跳上报与 LWT 遗嘱(设计文档 5.2/5.3/8 章)。
 *
 * topic: oc/devices/{device_id}/sys/property/report (QoS1, retain false)。
 * content 为探针节点数组, {device_id, ecuCode, nodeType} 三元组定位节点;
 * 周期 300s ±10% 抖动, MQTT CONNECT 后 5s 内全量, 节点上下线/ruleVersion
 * 变化即时增量上报。
 *
 * LWT(5.3): CONNECT 携带 will, 同通道, payload 为 nodeStatus=0 单节点数组,
 * broker 代发置离线。
 */
object PropertyReport {

    data class NodeProperty(
        val ecuCode: String,
        val ecuOs: String = "Android",
        val nodeType: String,
        val nodeVersion: String,
        val nodeStatus: Int = 1,   /* 0 离线 / 1 在线 / -1 异常 */
        val ruleVersion: String,
        val remark: String = "",
    )

    fun buildReport(nodes: List<NodeProperty>, timestampMs: Long): String {
        val content = JSONArray()
        for (n in nodes) {
            content.put(JSONObject().apply {
                put("ecuCode", n.ecuCode)
                put("ecuOs", n.ecuOs)
                put("nodeType", n.nodeType)
                put("nodeVersion", n.nodeVersion)
                put("nodeStatus", n.nodeStatus)
                put("ruleVersion", n.ruleVersion)
                if (n.remark.isNotEmpty()) put("remark", n.remark)
            })
        }
        return JSONObject().apply {
            put("msg_type", "property")
            put("protocol_version", "1.0")
            put("timestamp", timestampMs)
            put("manufacturer", DeviceIdentity.manufacturer)
            put("content", content)
        }.toString()
    }

    /** LWT 载荷: nodeStatus=0 单节点数组(5.3) */
    fun buildLwtPayload(primaryNodeType: String): String =
        buildReport(
            listOf(
                NodeProperty(
                    ecuCode = DeviceIdentity.ecuCode(),
                    nodeType = primaryNodeType,
                    nodeVersion = "idsm_manager_apk/1.0.0",
                    nodeStatus = 0,
                    ruleVersion = "unknown",
                )
            ),
            timestampMs = System.currentTimeMillis(),
        )

    /** 下一次全量上报时刻: base + 300s ± 10% 抖动(5.2) */
    fun nextReportTimeMs(nowMs: Long): Long {
        val jitter = Random.nextLong(-30_000, 30_001)
        return nowMs + 300_000 + jitter
    }
}

File: linux/idsm_managerd/property_report.cpp
/*
 * property_report.cpp -- 设备属性/心跳上报与 LWT 遗嘱(5.2/5.3/8 章)。
 */
#include "property_report.h"

#include <nlohmann/json.hpp>

#include <cstdlib>

namespace idsm {

std::string buildPropertyReport(const std::string& manufacturer,
                                const std::vector<NodeProperty>& nodes,
                                long long timestamp_ms) {
    nlohmann::json content = nlohmann::json::array();
    for (const auto& n : nodes) {
        nlohmann::json e;
        e["ecuCode"] = n.ecu_code;
        e["ecuOs"] = n.ecu_os;
        e["nodeType"] = n.node_type;
        e["nodeVersion"] = n.node_version;
        e["nodeStatus"] = n.node_status;
        e["ruleVersion"] = n.rule_version;
        if (!n.remark.empty()) e["remark"] = n.remark;
        content.push_back(std::move(e));
    }
    nlohmann::json env;
    env["msg_type"] = "property";
    env["protocol_version"] = "1.0";
    env["timestamp"] = timestamp_ms;
    env["manufacturer"] = manufacturer;
    env["content"] = std::move(content);
    return env.dump();
}

std::string propertyTopic(const std::string& device_id) {
    return "oc/devices/" + device_id + "/sys/property/report";
}

long long nextReportTimeMs(long long now_ms, unsigned random_jitter_ms) {
    /* 300s ± 10% 抖动; 抖动由调用方用 rand 或 urandom 生成 */
    const long long base = 300000;
    const long long half = base / 10;
    const long long jitter =
        random_jitter_ms % static_cast<unsigned>(2 * half + 1);
    return now_ms + base - half + jitter;
}

}  /* namespace idsm */
